### PR TITLE
Prevent going to profile page on valid auth page

### PR DESCRIPTION
**Pull request by sionking**
if no auth that recursivly trigger the auth method on profile page

### CODE REVIEW

**Comment by timgrossmann** [comment]
@sionking Do we ever have `stop=True`?
Or more like somewhere in the code at the moment?

Thank you for your contribution.

**Comment by sionking** [comment]
This is the stoping condition for the recursive check ... we have a stop=True inside the function it self.

**Comment by timgrossmann** [comment]
@sionking Ahh there we go, wasn't expecting a recursive call.
Thank you for the clarification.

File: instapy/util.py [hunk 1]
```diff
@@ -1056,15 +1056,15 @@ def load_user_id(username, person, logger, logfolder):
...
+def check_authorization(browser, username, method, logger, stop=False):
     """ Check if user is NOW logged in """
     logger.info("Checking if '{}' is logged in...".format(username))
 
     # different methods can be added in future
     if method == "activity counts":
 
         profile_link = 'https://www.instagram.com/{}/'.format(username)
-        web_address_navigator(browser, profile_link)
+        #web_address_navigator(browser, profile_link)
```
Comment (timgrossmann) [inline instapy/util.py:1067]: Could we remove this instead of commenting it out, please?

```suggestion
```

**Comment by uluQulu** [comment]
Hi there
I think you have not seen the commits in my live PR.
I had that part written already at https://github.com/timgrossmann/InstaPy/pull/3318/commits/b0725e1af9634323db1c61af36c03a94e479fa02 commit.

The idea belongs to @sionking as he reminded me why we need that extra navigations in some thread.
So I wrote it while adding `check_authorization` function to be used at **login_util.py** file to _speed up_ the `logging procedure` (_see https://github.com/timgrossmann/InstaPy/pull/3318/commits/ee6acba1f43f71083a24aa32290f9a6247244859 commit_)..

I saw that @sionking uses `stop`, `auth` stuff;
Well, it is error prone and inefficient.

I used the logic of,
- If it is stucked on an _abnormal_ webpage, navigate to _a normal_ webpage (_e.g. logged in user's profile page_).
_It pays nothing from the efficiency_.

---

Also, it's me who added that navigation to the profile page, and is better to remind again,
I did it cos in some cases **InstaPy** can be stucked at an abnormal webpage, more specifically, e.g. while getting the username from the user ID, it navigates to an unusual IG page, etc.
For such reasons, I did that [extra] navigation to make sure that checking the authorization will take place properly.

---

Cheers 😁
---

**Comment by sionking** [comment]
> Hi there
> I think you have not seen the commits in my live PR.
> I had that part written already at [b0725e1](https://github.com/timgrossmann/InstaPy/commit/b0725e1af9634323db1c61af36c03a94e479fa02) commit.
> 
> The idea belongs to @sionking as he reminded me why we need that extra navigations in some thread.
> So I wrote it while adding `check_authorization` function to be used at **login_util.py** file to _speed up_ the `logging procedure` (_see [ee6acba](https://github.com/timgrossmann/InstaPy/commit/ee6acba1f43f71083a24aa32290f9a6247244859) commit_)..
> 
> I saw that @sionking uses `stop`, `auth` stuff;
> Well, it is error prone and inefficient.
> 
> I used the logic of,
> 
> * If it is stucked on an _abnormal_ webpage, navigate to _a normal_ webpage (_e.g. logged in user's profile page_).
>   _It pays nothing from the efficiency_.
> 
> Also, it's me who added that navigation to the profile page, and is better to remind again,
> I did it cos in some cases **InstaPy** can be stucked at an abnormal webpage, more specifically, e.g. while getting the username from the user ID, it navigates to an unusual IG page, etc.
> For such reasons, I did that [extra] navigation to make sure that checking the authorization will take place properly.
> 
> ## Cheers 😁

God why are you writing so many lines! 
Just have a look again, I will navigate to the profile page in case we can't find auth in page we are currently in.
* I also think your PR can fit with this method as an extra check.

**Comment by uluQulu** [comment]
@sionking 
>God why are you writing so many lines!

I have always wanted to cover all aspects of a subject.
But you're right. I must learn to write shorter 🙁 as possible 😀

This will also be a long post but consider that you may be aware of some words I told out of the conversation we had in other threads but e.g. @timgrossmann or any other reader of this PR is not aware of those talks and this merging some code is a serious step that needs everything be considered.

>Just have a look again, I will navigate to the profile page in case we can't find auth in page we are currently in.

I have looked, mate. Your approach is inefficient in many ways.
Regardless of the current URL whether it is normal or abnormal, you always try to find your element which is a total waste.
Also, if user really did not log in for some reason and the page is really valid, then you still navigate to profile page and re-run same logic = extra waste.

>I also think your PR can fit with this method as an extra check.

It can fit, but my approach has not any leak to need this.
And in my approach, the only extra time spent is just checking the URL of the current page for 1 time which consumes averagely `0.015007734298706055` seconds [in my slow connection].

#### In other words


You go to collect watermelons from a lemon yard and after finishing, realize that all you have collected is lemons, but not watermelons and return back empty with lots of time & lemons wasted and only then navigate to the watermelons yard.




I go to collect watermelons from a lemon yard but before collecting I do run `get_current_yard()` which tells me that I'm currently in a lemon yard, so that I directly navigate to watermelon's home yard.



---



I don't know there is lots of to say about this little logic, just thinking about the consequences is enough.
Sorry for another long post, @sionking 😟

---

Cheers 😁
---

**Comment by instasck** [comment]
@uluQulu you missed one thing.
Statistics.
Statistically the method failure is 0.001%, So no time is wasted here.

**Comment by uluQulu** [comment]
@instasck,
I live in a town and saw many watermelon yards and I can assure you, when watermelons are small, they are much more green and does not look yellowish.
Also, lemons do grow on trees.

I suggest, go and investigate it. You might have eye vision deficiency.




---

---


>...Statistically the method failure is 0.001%,...

It is error prone, if so, percentage is unimportant.

>...So no time is wasted here....

It is inefficient. I have explained 'why' in the posts above about those "wastes".

But don't worry, guys. I will no more do analyze your PRs 🚥
---
Is that your other account, @sionking? (_just a joke_)

🌑
---

**Comment by sionking** [comment]
@uluQulu you missed one thing.
Statistics.
Statistically the method failure is 0.001%, So no time is wasted here.

Sorry, one more thing, your method will fail on lemon yard that you think are lemons but are just a small watermelons!

**Comment by sionking** [comment]
@uluQulu Yes this is me from another account (oops).


> It is error prone, if so, percentage is unimportant.

Haa so arrogant! take one example for it to fail;
While most of you cases relay on **current** implementation of Instapy, I think a bit wider for general case. So my case was when I "story watch" with the bot, I did not get any auth == True while I was logged in. what did happen is that the story got stuck in loading, and the app took it as a logged out.

So here's an example for **must** have this extra test.
BTW, even if you check the url you will see instagram.com in it.

image

**Comment by timgrossmann** [comment]
@sionking @uluQulu Guys, even though I think your conversations are always fun to read, please keep in mind the [Code of Conduct](https://github.com/timgrossmann/InstaPy/blob/master/CODE_OF_CONDUCT.md) we have here at this project.
We don't want people to feel intimidated or not welcome. Maybe conversations like this will make new developers think twice and hesitate before creating a PR.

Thanks for understanding this 😉 👌

**Comment by sionking** [comment]
Please merge

**Comment by timgrossmann** [comment]
@sionking Please resolve the merge conflicts before we merge this.
Thank you

@sionking @uluQulu Also I'm still not absolutely sure what this arguing was about, is this working (with the merged PR of @uluQulu) and ready to be merged?

**Comment by sionking** [comment]
I think if @uluQulu can't disprove my scenario, it will be best to merge it.

**Comment by uluQulu** [comment]
I don't think I can disprove something, @sionking but I will reply.
But now, it makes some sense 🙂
>"... story got stuck in loading, and the app took it as a logged out."

You mean you can't query the server to get activity counts data during that lag,. weird.
It is a bit frustrating to produce that situation, so I will assume your situation to be real and unavoidable;

>BTW there is no any feature related to "Stories" at the moment.
So, you shan't expect me [or others] to think of situations from unexisted features.


Here is how to do it efficiently [in plain words],
```python
URL = get_current_url

If not URL or URL is abnormal:
    navigate to a valid page [e.g. profile page];

get "activity counts";

# EXTRA [for your weird "Story" situation]:
If URL is normal and authorization failed (no "activity counts"):
    force-navigate to a valid page [e.g. profile page];
    get "activity counts";

```
It is 1 navigation in all possible costs.

One of the difference of it from your approach is,
+ Your code will try to get activity counts data on an abnormal page which is a total
waste;
+ My approach makes sure the page is normal before trying to get activity counts data;

Also, recursive function here is an overuse and a bit unfit?
E.g. [in case of an authorization failure] your code will enter to the `if not auth` block twice and duplicate print the logger message in there.

It is best to move the process of getting activity counts into a separate function rather than using recursion.

---

It is error prone, cos if some contributor behaves that formal argument differently, it will fail.

---


Solutions come out of needs.
Needs raise by time.
You can't know the future.

It's like no car has been invented yet and you are making a wheel to be used in future 😂 for that mystic 'car' ...

---

I know it is not best place to tell it but as recursion came on spot, I could say that
I work on a new mind-blowing feature called nested interactions which uses recursion
like nothing 🌌

Sorry, @timgrossmann, I am just imperfect.
@sionking, throughout this project, I have learnt lots of priceless knowledge from you from day 1, so in any case I am thankful to you.


---
I have never been a **Google** fan, earlier **Bing**, I was using **DuckDuckGo** on **Firefox**.
But I choose **Google** sometimes e.g., to get more _Stackoverflow_ results (_they have a deal on it 🤝🏼?_) and all other times **Bing** has more relevant results, smarter interface, easier photo download, beautiful looks and is supported well by my general browser of choice- **Edge**.

Cheers 😁
---

<sub>[**Bing**](https://www.bing.com) is good, use it 🥞☕</sub>

**Comment by sionking** [comment]
@uluQulu  again endless story  :) aren't you married/kids ?


> It is a bit frustrating to produce that situation, so I will assume your situation to be real and unavoidable;

Any time you are in a story you will get into it.



> Also, recursive function here is an overuse and a bit unfit?

Fit perfectly IMO. Plus your addition to "normal" page ++
No extra code only one stop if condition... No need to check activity count again. 
You are right that the recursive is only 2 levels but still it is working and efficient.

BTW, we are not NASA here, if something is working no need to improve it to perfect condition
We have much of other work to do !

> You can't know the future.

Some religious says they can... and also good engineering is to predict and design not only current plan. 


@uluQulu I also learn alot from you ! really I like your code style (most of the time) :)

**Comment by timgrossmann** [comment]
@sionking Could you please resolve the conflicts so we can merge this?
Please also add a new entry to the CHANGELOG file.

Thanks guys 👍

**Comment by uluQulu** [comment]
@sionking 

>@uluQulu again endless story :) aren't you married/kids ?

No. Also, it's not a reason to lose professionalism.
 
>Fit perfectly IMO. .. 
No extra code only one stop if condition... 

That [extra] code is to handle your ["Story" feature] situation in its best.
Moreover, "_only one stop if condition_" - recursion is a way complex to understand than my straight-forward approach.
I'd suggest to avoid recursion unless truly necessary.

>No need to check activity count again.

You already check activity counts twice [in failures] and it has a redundant check in abnormal pages.

>You are right that the recursive is only 2 levels but still it is working and efficient.

My prior opinions are the same on it.

>BTW, we are not NASA here, if something is working no need to improve it to perfect condition
We have much of other work to do !

I understand you, not everyone can be this detail specific for an open-source free application. But for me, code is code- it must be good.
And the reason I joined this PR is you did what I already did.


>Some religious says they can... and also good engineering is to predict and design not only current plan.

I respect your beliefs.
But prediction is an illusional act. Let's be precise, you can't know until it happens 🤔 can you?

---


Cheers 😁
---

<sub>Enjoy the moment 🚀</sub>
